pkg/apis/volumesnapshot/v1alpha1/types.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+
+	core_v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	// VolumeSnapshotDataResourcePlural is "volumesnapshotdatas"
+	VolumeSnapshotDataResourcePlural = "volumesnapshotdatas"
+	// VolumeSnapshotResourcePlural is "volumesnapshots"
+	VolumeSnapshotResourcePlural = "volumesnapshots"
+	// SnapshotClassResourcePlural is "snapshotclasss"
+	SnapshotClassResourcePlural = "snapshotclasses"
+)
+
+// VolumeSnapshotStatus is the status of the VolumeSnapshot
+type VolumeSnapshotStatus struct {
+	// The time the snapshot was successfully created
+	// +optional
+	CreationTimestamp metav1.Time `json:"creationTimestamp" protobuf:"bytes,1,opt,name=creationTimestamp"`
+
+	// Represent the latest available observations about the volume snapshot
+	Conditions []VolumeSnapshotCondition `json:"conditions" protobuf:"bytes,2,rep,name=conditions"`
+}
+
+// VolumeSnapshotConditionType is the type of VolumeSnapshot conditions
+type VolumeSnapshotConditionType string
+
+// These are valid conditions of a volume snapshot.
+const (
+	// VolumeSnapshotConditionCreating means the snapshot is being created but
+	// it is not cut yet.
+	VolumeSnapshotConditionCreating VolumeSnapshotConditionType = "Creating"
+	// VolumeSnapshotConditionUploading means the snapshot is cut and the application
+	// can resume accessing data if ConditionStatus is True. It corresponds
+	// to "Uploading" in GCE PD or "Pending" in AWS and ConditionStatus is True.
+	// This condition type is not applicable in OpenStack Cinder.
```

Comment:
maybe make this comment a little more generic. "Uploading" condition is only applicable to the volume plugins which support uploading snapshots to Cloud storage. For example, it corresponds to snapshot "UPLOADING" status in GCE and ...

Reply:
I didn't see the change?

Reply:
I read the updated condition guideline again, and have the following thoughts.

Condition should be considered as observations that users cared about instead of state-machine phases. For snapshot, user cares about whether snapshot is created successfully or not first. These can be represented by "ready" and "failed" conditions. The other condition user cares about is when the snapshot is taken (or cut). Because some plugins support uploading which might take a long time. User needs to know when the snapshot is taken (usually in a second) so that application can be resumed. So we could add another condition "snapshotTaken" (or just "Taken") condition. So now we have three conditions "snapshotTaken", "snapshotReady" and "snapshotFailed"

During the snapshot creation, we might see the following conditions at different times.
1. First when user create VolumeSnapshot object, there is no condition yet. (means controller is working on it, but there is no observed state available yet) 
2. The volume plugin starts to create the snapshot. When the snapshot is taken (cut), controller will get a response and add "SnapshotTaken" condition to true. If snapshot is failed, it will be set to false. If there is no uploading phase, Ready condition will also be true right after snapshot is taken.
3. If there is an uploading phase, the controller will add "Ready" condition to true after uploading finishes. If uploading failed, the condition is set to false. 

Please let me know if you have any feedback/comment about it. Thanks!